earthkit/data/decorators.py

```diff
@@ -35,6 +35,19 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def detect_out_filename(func):
+    @functools.wraps(func)
+    def wrapped(self, *args, **kwargs):
+        if len(args) == 0:
+            try:
+                args = [os.path.basename(self.path)]
```

Comment:
What if I have this code:

```python
ds = from_source("file", "my.grib")
ds.save()
```
or
```python
ds = from_source("file", "my.grib")
ds.save(append=True)
```

Reply:
<img width="785" alt="image" src="https://github.com/ecmwf/earthkit-data/assets/53045993/e7f9288d-eb2b-4b44-90d2-be9f4f3fd6b7">

Reply:
ds.save() created a file called test6.grib in the current directory.

the `append` kwarg was not recognised with the decorator... just looking into why

Reply:
The first example overwrites the original file we just read with `from_source()`!!!
```python
ds = from_source("file", "my.grib")
ds.save()
```

The second the example would append to itself to the original file:
```python
ds = from_source("file", "my.grib")
ds.save(append=True)
```

Neither should happen!

Reply:
but doing this would produce the same result in the current implementation:
```
ds = from_source("file", "my.grib")
ds.save("my.grib")
ds.save("my.grib", append=True)
```

Reply:
In these examples `ds` is a not FileSource!

Reply:
GribReader, sorry.

Reply:
Actually that didn't fix the exisiting bug that it is possible to overwrite the file you are reading. The latest commit fixes this with a warning

Reply:
> GribReader, sorry.

Still the error says `TypeError` FileSource.save()` so something goes wrong.

Reply:
The error I posted above? In the end I realised that that was caused by what we have been discussing, when we saved we tried to overwrite the file we were reading, and this created a broken grib, so when I opened it again it was not understood as a grib. This is fixed by the check to not overwrite the existing file, but I am going to look into the preserving the source filename before re-opening this ticket.

Reply:
This point has been addressed with the latest changes. I have added tests that ensure that the warning is raised when the user attempts to overwrite the file they are reading.